### ISSUE TITLE
Why not causal-masking MLP layer?

### ISSUE THREAD

**Issue by MikeTkachuk** [karpathy/minGPT#139]
upd:
Messed up the shapes, it's fine

Current flow: 
  input sequence: x
  ugly output: 
```
 x + att1(x) + mlp1(ln(x+att1(x))) +   # block 1
att2(x + att1(x) + mlp1(ln(x+att1(x)))) + \
  mlp2(ln( x + att1(x) + mlp1(ln(x+att1(x))) + att2(x + att1(x) + mlp1(ln(x+att1(x)))))) + ...  # block 2
```
As a model I can learn the following flow:
  output: 
`x + 0 + mlp1(ln(x + 0)) + 0 + ... + 0`
Now since we are not masking future tokens from the mlp1's linear layer, I can learn something like this:
```
"""
-1 1 0 0 ...
0 -1 1 0 ...
0 0 -1 1 ...
0 0 0 -1 ...
...
"""
mlp1 = torch.diag(torch.ones(), diagonal=1) - torch.diag(torch.ones(), diagonal=0)
```
So that `x + x @ mlp1 = x - x + x[1:]`

Is this desired?
